### PR TITLE
Refactor library

### PR DESCRIPTION
**Pull request by kantai**
This PR:

1. Adds eslint definition, and lints the codebase
2. Adds flow types.
3. Fixes `for-in` usage which the flow typing reveals.
4. Adds circleCI definition.

### CODE REVIEW

File: src/makeZoneFile.js [hunk 1]
```diff
@@ -0,0 +1,191 @@
+/* @flow */
```
Comment (zone117x) [inline src/makeZoneFile.js:1]: I hate how bad git is at detecting file name changes :(

**Review by zone117x — APPROVED** [review]
Do we want to add a unit test for the bug this fixes? This may be a good example of where we can draw some QA lines - as in what parameters about this situation make us not care about adding a regression test or something like that?

Other than that and assuming the renamed files (that git diff shows as deletes and additions) are just syntax changes then looks 👍

File: src/makeZoneFile.js (hunk 1 above)
Comment (kantai) [reply src/makeZoneFile.js:1]: Yeah, agreed.

**Comment by kantai** [comment]
Thanks @zone117x !

I added a regression test for the behavior experienced when used with Ember (see `testZoneFileMakingFromExtendedArray`)

Also, added a test for behaviors related to using zone file templates with "unmatched" template fields, which also revealed a regression in the refactor.

**Pull request merged**